From 9afd9edd87b644e42e1c8841126777d2dc15ba26 Mon Sep 17 00:00:00 2001
From: Jinzhe Zeng <jinzhe.zeng@ustc.edu.cn>
Date: Tue, 4 Nov 2025 22:19:42 +0800
Subject: [PATCH] fix(cmake): improve CUDA C++ standard setting for
 compatibility with gcc-14

Signed-off-by: Jinzhe Zeng <jinzhe.zeng@ustc.edu.cn>
(cherry picked from commit d1986841decc01b50e7bf20ac677271bebcb438c)
---
 source/lib/src/gpu/CMakeLists.txt | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/source/lib/src/gpu/CMakeLists.txt b/source/lib/src/gpu/CMakeLists.txt
index 0d176dc32..e622f3486 100644
--- a/source/lib/src/gpu/CMakeLists.txt
+++ b/source/lib/src/gpu/CMakeLists.txt
@@ -42,6 +42,12 @@ if(USE_CUDA_TOOLKIT)
     message(FATAL_ERROR "CUDA version must be >= 9.0")
   endif()
 
+  # NVCC compilation errors with gcc-14 and c++11 Cases in other repos:
+  # https://gitlab.archlinux.org/archlinux/packaging/packages/cuda/-/issues/12
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "14")
+    set_if_higher(CMAKE_CUDA_STANDARD 14)
+  endif()
+
   message(STATUS "NVCC version is " ${CMAKE_CUDA_COMPILER_VERSION})
 
   # arch will be configured by CMAKE_CUDA_ARCHITECTURES
-- 
2.34.1

